pyro/ops/stats.py

```diff
@@ -0,0 +1,160 @@
+from __future__ import absolute_import, division, print_function
+
+import torch
+
+
+def _compute_chain_variance_stats(input):
+    # compute within-chain variance and variance estimator
+    N = input.size(1)
+    chain_mean = input.mean(dim=1)
+    var_between = chain_mean.var(dim=0)
+
+    chain_var = input.var(dim=1)
+    var_within = chain_var.mean(dim=0)
+    var_estimator = (N - 1) / N * var_within + var_between
+    return var_within, var_estimator
+
+
+def gelman_rubin(input):
+    """
+    Computes R-hat over chains of samples. The first two dimensions of `input`
+    is `C x N` where `C` is the number of chains and `N` is the number of samples.
```

Comment:
Not part of this PR, but we will need to refactor our `MCMC` (and, even `TracePosterior`) classes to aggregate samples by chain. I will start thinking about that.

Reply:
In that case, it will be good to have two arguments `chain_dim`, `sample_dim`. :D

Reply:
Sounds good! If it's too much of an effort to generalize, I think the choice of `chain_dim` and `sample_dim` is fine. This is anyways what would have been the default convention to have chain be leftmost followed by samples.

Reply:
In this case, I have to add two dummy dimensions at the front of input, then transpose sample and chain dimensions to these two dummies. Finally, I have to squeeze them. This way it is easier to follow the code without worrying about the positions of dimensions.